src/default_storage.jl

```diff
@@ -0,0 +1,393 @@
+    # be heap allocated, so there's no reason to strongly type their dict. But inputs can
+    # be isbits, so it's probably worth specailizing them.
+    inputs_map::Dict{Tuple{Type,Tuple},InputValue}
+    derived_function_maps::DerivedFunctionMapType
+
+    # Tracks whether there are any derived functions currently active. It is an error to
+    # modify any inputs while derived functions are active, on the current Task or any Task.
+    derived_functions_active::Int
+
+    function DefaultStorage()
+        new(Base.ReentrantLock(), 0, InputMapType(), DerivedFunctionMapType(), 0)
+    end
+end
+
+const DefaultRuntime = Salsa.Runtime{Salsa.EmptyContext,DefaultStorage}
+
+function Base.show(io::IO, storage::DefaultStorage)
+    current_revision = lock(storage.lock) do
+        storage.current_revision
+    end
+    print(io, "Salsa.DefaultStorage($current_revision, ...)")
+end
+
+
+# NOTE: This implements the dynamic behavior for Salsa Components, allowing users to define
+# input/derived function dynamically, by attaching new Dicts for them to the storage at
+# runtime.
+function get_map_for_key(storage::DefaultStorage, key::DerivedKey{<:Any,TT}) where {TT}
+    try
+        lock(storage.lock)
+        return get!(storage.derived_function_maps, key) do
+            # PERFORMANCE NOTE: Only construct key inside this do-block to
+            # ensure expensive constructor only called once, the first time.
+
+            # TODO: Use the macro's returntype to strongly type the value.
+            #       We'll have to generate this function from within the macro, like we used to
+            #       in the existing open-source Salsa.
+            # NOTE: Except actually after https://github.com/RelationalAI-oss/Salsa.jl/issues/11
+            #       maybe we won't do this anymore, and we'll just use one big dictionary!
+            Dict{TT,DerivedValue}()
+        end
+    finally
+        unlock(storage.lock)
+    end
+end
+function get_map_for_key(storage::DefaultStorage, ::InputKey)
+    # We use one big dictionary for the inputs, storing them all together, so any input key
+    # would return the same value here. :)
+    return storage.inputs_map
+end
+
+
+function Salsa._previous_output_internal(
+    runtime::Salsa._TracingRuntimeWithStorage{DefaultStorage},
+    key::DependencyKey{<:DerivedKey},
+)
+    storage = _storage(runtime)
+    derived_key, args = key.key, key.args
+
+    previous_output = nothing
+
+    try
+        lock(storage.lock)
+        cache = get_map_for_key(storage, derived_key)
+
+        if haskey(cache, args)
+            previous_output = getindex(cache, args)
+        end
+    finally
+        unlock(storage.lock)
+    end
+
+    return previous_output
+end
+
+
+# TODO: I think we can @nospecialize the arguments for compiler performance?
+# TODO: It doesn't seem like this @nospecialize is working... It still seems to be compiling
+# a nospecialization for every argument type. :(
```

Comment:
Filed https://github.com/RelationalAI-oss/Salsa.jl/issues/19 for the nospecialize comments throughout this PR